### ISSUE TITLE
Can't get multiple instances to connect (Ubuntu 21.04)

### ISSUE THREAD

**Issue by DiamondDust132** [hrydgard/ppsspp#14906]
### Game or games this happens in

Every game I've tried so far...

### What area of the game / PPSSPP

Hello. I am unable to get multiplayer working from the same PC. I've had no issues with public servers or VPNs like ZeroTier when it comes to playing online with others, but I just can't get local multiplayer working. I've poured over [this](https://github.com/hrydgard/ppsspp/wiki/How-to-play-multiplayer-games-with-PPSSPP) tutorial dozens of times to no avail. The first instance seems to work properly as I see 'Network Initialized' and can successfully get into a lobby. The second instance of PPSSPP simply doesn't connect to the first and gets stuck on a looping connecting screen or simply doesn't connect at all. I've set the IP to 'localhost' on both instances, made sure only one instance is the host by checking 'Enable built-in PRO ad hoc server,' tried changing the WLAN channel from 'Auto' to something else, made sure both instances had the same port offset.

![Screenshot from 2021-09-21 08-19-52](https://user-images.githubusercontent.com/20025151/134169308-0a7115b5-f710-4aa1-a039-c87651ef5f74.png)
First Instance Working like it should

![Screenshot from 2021-09-21 08-19-55](https://user-images.githubusercontent.com/20025151/134169341-5f476dce-ee2d-46ba-bd5c-dfe3eeb64ba8.png)
Second instance being stuck in a connection loop I'm presuming. It just stays on this screen forever.

![Screenshot from 2021-09-21 08-23-36](https://user-images.githubusercontent.com/20025151/134169962-c131fdb5-d184-4051-bd57-df0ce8f4d538.png)
Network settings of the host

![Screenshot from 2021-09-21 08-23-38](https://user-images.githubusercontent.com/20025151/134169920-0d4aa9a4-c3a8-4656-8345-12039fc58cc9.png)
Network settings of the same instance (Port offset is the same for both: 10000)

It's not just Bleach: Heat the Soul 7 having this problem. No games I've tried work. Is it perhaps just a Linux issue? I didn't see anywhere in the tutorial though that this wasn't a multiplatform feature. Unlike Windows, I'm not given an option to 'Open new instance' from any sort of menu. I simply open another instance of PPSSPP manually, which the tutorial says is alright. I've tried old versions of the emulator to no avail. Any help would be appreciated because this is honestly driving me crazy. I know Bleach: Heat the Soul 7 in particular works with this method, but I just can't wrap my head around this.

### What should happen

Both instances should connect to allow for local multiplayer.

### Platform

Linux / BSD

### Mobile phone model or graphics card

Radeon RX 570

### PPSSPP version affected

v1.11.3-1517-gb8ea95bec

### Graphics backend (3D API)

OpenGL / GLES

### Checklist

- [X] Test in the [latest git build](https://buildbot.orphis.net/ppsspp/) in case it's already fixed.
- [X] [Search for other reports](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same issue.
- [X] Try resetting settings or older versions and include if the issue is related.
- [X] Try without any cheats and without loading any save states.
- [X] Include logs or screenshots of issue.

**Comment by ANR2ME** [comment]
Could you provide the debug logs of both instances? you will need to build the debug version to generate the debug logs on non-windows i think.

**Comment by MojoJojoDojo** [comment]
I think that instead of localhost you will need to use another IP like 127.0.0.1 and 127.0.0.2 on the other instance and so on.
I know that worked before at least on Windows as well.

**Comment by ANR2ME** [comment]
> 
> 
> I think that instead of localhost you will need to use another IP like 127.0.0.1 and 127.0.0.2 on the other instance and so on.
> I know that worked before at least on Windows as well.

PPSSPP should be able to translate "localhost" into the correct 127.0.0.x IP automatically just like on Windows, otherwise there is a bug on it.

**Comment by iota97** [comment]
Multiple instance detection was broken indeed, just to note this error log doesn't seem to work
https://github.com/hrydgard/ppsspp/blob/b8ea95bec1da04b75ebf295ea581a8f7e01dd472/Core/Instance.cpp#L152

I think our log manger was not ready yet or something.

**Comment by DiamondDust132** [comment]
Knew I forgot to add something to my post. Ugh. Sorry about that. Here are the logs. Both are identical actually other than the mad address, so I don't think PPSSPP sees them as a part of the same instance? Like what MojoJojoDojo was saying, I believe. Both localhosts end up as 127.0.0.1 even if I change the mac address on the second instant.

**Log for 1st Instance**
VulkanMayBeAvailable: Device allowed ('SDL:Linux')
VulkanMayBeAvailable: Library loaded ('libvulkan.so')
VulkanMayBeAvailable: Enumerating instance extensions
VulkanMayBeAvailable: Instance extension count: 18
VulkanMaybeAvailable: Instance extension found: VK_KHR_device_group_creation (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_display (00000017)
VulkanMaybeAvailable: Instance extension found: VK_KHR_external_fence_capabilities (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_external_memory_capabilities (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_external_semaphore_capabilities (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_get_display_properties2 (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_get_physical_device_properties2 (00000002)
VulkanMaybeAvailable: Instance extension found: VK_KHR_get_surface_capabilities2 (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_surface (00000019)
VulkanMaybeAvailable: Instance extension found: VK_KHR_surface_protected_capabilities (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_wayland_surface (00000006)
VulkanMaybeAvailable: Instance extension found: VK_KHR_xcb_surface (00000006)
VulkanMaybeAvailable: Instance extension found: VK_KHR_xlib_surface (00000006)
VulkanMaybeAvailable: Instance extension found: VK_EXT_acquire_xlib_display (00000001)
VulkanMaybeAvailable: Instance extension found: VK_EXT_debug_report (0000000a)
VulkanMaybeAvailable: Instance extension found: VK_EXT_direct_mode_display (00000001)
VulkanMaybeAvailable: Instance extension found: VK_EXT_display_surface_counter (00000001)
VulkanMaybeAvailable: Instance extension found: VK_EXT_debug_utils (00000002)
VulkanMayBeAvailable: Calling vkCreateInstance
VulkanMayBeAvailable: Vulkan test instance created successfully.
VulkanMayBeAvailable: Ineligible device found and ignored: 'llvmpipe (LLVM 12.0.0, 256 bits)'
VulkanMayBeAvailable: Found working Vulkan API!
VulkanMayBeAvailable: Destroying instance
DEBUG: Vulkan might be available.
Info: We compiled against SDL version 2.0.12, but we are linking against SDL version 2.0.14., be aware that this can lead to unexpected behaviors
ThreadManager::Init(compute threads: 16, all: 20)
16:54:524 Core/Config.cpp:619 I[G3D]: Longest display side: -1 pixels. Choosing scale 1
Pixels: 960 x 544
Virtual pixels: 960 x 544
OpenGL 2.0 or higher.
loading control pad mappings from gamecontrollerdb.txt: SUCCESS!
Control pad device 0 not supported by SDL game controller database, attempting to create default mapping...
Added default mapping ok
found control pad: SWITCH CO.LTD. USB Gamepad, loading mapping: SUCCESS, mapping is:
03000000790000001100000011010000,SWITCH CO.LTD. USB Gamepad,x:b3,a:b0,b:b1,y:b2,back:b8,guide:b10,start:b9,dpleft:b15,dpdown:b14,dpright:b16,dpup:b13,leftshoulder:b4,lefttrigger:a2,rightshoulder:b6,rightshoulder:b5,righttrigger:a5,leftstick:b7,leftstick:b11,rightstick:b12,leftx:a0,lefty:a1,rightx:a3,righty:a4
pad 1 has been assigned to control pad: SWITCH CO.LTD. USB Gamepad
16:54:826 Net/NetBuffer.cpp:41 E[IO]: FlushSocket failed
16:54:826 Net/NetBuffer.cpp:41 E[IO]: FlushSocket failed
17:56:470 HLE/sceNet.cpp:213 N[SCENET]: LocalHost IP will be 127.0.0.1 [4b:53:4f:8f:41:fb]
17:56:574 root         N[G3D]: GLES/ShaderManagerGLES.cpp:910 Precompiling the shader cache from '/home/aaron/.config/ppsspp/PSP/SYSTEM/CACHE/UCJS10110.glshadercache'
17:56:574 root         N[G3D]: GLES/ShaderManagerGLES.cpp:990 Precompile: Compiled and linked 35 programs (15 vertex, 16 fragment) in 0.3 milliseconds
17:56:574 root         N[BOOT]: UI/EmuScreen.cpp:341 Loading /home/aaron/Downloads/ppsspp_old/Bleach_Heat_The_Soul_7_JPN_PSP-Caravan.iso...
18:02:235 Main         E[SCEIO]: HLE/sceIo.cpp:2450 Not a valid PGD file. Open as normal file.
18:02:235 Main         E[HLE]: HLE/HLE.cpp:374 sceNpDrmEdataSetupKey: Delaying a thread that's already waiting
18:02:267 Main         E[SCEIO]: HLE/sceIo.cpp:2450 Not a valid PGD file. Open as normal file.
18:02:267 Main         E[HLE]: HLE/HLE.cpp:374 sceNpDrmEdataSetupKey: Delaying a thread that's already waiting
18:36:352 Main         E[SCEKERNEL]: HLE/sceKernelThread.cpp:1340 80020198=sceKernelGetThreadExitStatus(0)

**Log for 2nd Instance**
VulkanMayBeAvailable: Device allowed ('SDL:Linux')
VulkanMayBeAvailable: Library loaded ('libvulkan.so')
VulkanMayBeAvailable: Enumerating instance extensions
VulkanMayBeAvailable: Instance extension count: 18
VulkanMaybeAvailable: Instance extension found: VK_KHR_device_group_creation (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_display (00000017)
VulkanMaybeAvailable: Instance extension found: VK_KHR_external_fence_capabilities (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_external_memory_capabilities (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_external_semaphore_capabilities (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_get_display_properties2 (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_get_physical_device_properties2 (00000002)
VulkanMaybeAvailable: Instance extension found: VK_KHR_get_surface_capabilities2 (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_surface (00000019)
VulkanMaybeAvailable: Instance extension found: VK_KHR_surface_protected_capabilities (00000001)
VulkanMaybeAvailable: Instance extension found: VK_KHR_wayland_surface (00000006)
VulkanMaybeAvailable: Instance extension found: VK_KHR_xcb_surface (00000006)
VulkanMaybeAvailable: Instance extension found: VK_KHR_xlib_surface (00000006)
VulkanMaybeAvailable: Instance extension found: VK_EXT_acquire_xlib_display (00000001)
VulkanMaybeAvailable: Instance extension found: VK_EXT_debug_report (0000000a)
VulkanMaybeAvailable: Instance extension found: VK_EXT_direct_mode_display (00000001)
VulkanMaybeAvailable: Instance extension found: VK_EXT_display_surface_counter (00000001)
VulkanMaybeAvailable: Instance extension found: VK_EXT_debug_utils (00000002)
VulkanMayBeAvailable: Calling vkCreateInstance
VulkanMayBeAvailable: Vulkan test instance created successfully.
VulkanMayBeAvailable: Ineligible device found and ignored: 'llvmpipe (LLVM 12.0.0, 256 bits)'
VulkanMayBeAvailable: Found working Vulkan API!
VulkanMayBeAvailable: Destroying instance
DEBUG: Vulkan might be available.
Info: We compiled against SDL version 2.0.12, but we are linking against SDL version 2.0.14., be aware that this can lead to unexpected behaviors
ThreadManager::Init(compute threads: 16, all: 20)
16:56:640 Core/Config.cpp:619 I[G3D]: Longest display side: -1 pixels. Choosing scale 1
16:56:647 Net/HTTPServer.cpp:244 E[IO]: Failed to bind to port 40767, error=98 - Bailing (ipv6)
16:56:647 Net/HTTPServer.cpp:194 E[IO]: Failed to bind to port 40767, error=98 - Bailing (ipv4)
Pixels: 960 x 544
Virtual pixels: 960 x 544
OpenGL 2.0 or higher.
loading control pad mappings from gamecontrollerdb.txt: SUCCESS!
Control pad device 0 not supported by SDL game controller database, attempting to create default mapping...
Added default mapping ok
found control pad: SWITCH CO.LTD. USB Gamepad, loading mapping: SUCCESS, mapping is:
03000000790000001100000011010000,SWITCH CO.LTD. USB Gamepad,x:b3,a:b0,b:b1,y:b2,back:b8,guide:b10,start:b9,dpleft:b15,dpdown:b14,dpright:b16,dpup:b13,leftshoulder:b4,lefttrigger:a2,rightshoulder:b6,rightshoulder:b5,righttrigger:a5,leftstick:b7,leftstick:b11,rightstick:b12,leftx:a0,lefty:a1,rightx:a3,righty:a4
pad 1 has been assigned to control pad: SWITCH CO.LTD. USB Gamepad
16:56:821 Net/NetBuffer.cpp:41 E[IO]: FlushSocket failed
16:56:821 Net/NetBuffer.cpp:41 E[IO]: FlushSocket failed
18:15:958 HLE/sceNet.cpp:213 N[SCENET]: LocalHost IP will be 127.0.0.1 [e4:c5:bf:3d:f4:ad]
18:16:028 root         N[G3D]: GLES/ShaderManagerGLES.cpp:910 Precompiling the shader cache from '/home/aaron/.config/ppsspp/PSP/SYSTEM/CACHE/UCJS10110.glshadercache'
18:16:029 root         N[G3D]: GLES/ShaderManagerGLES.cpp:990 Precompile: Compiled and linked 35 programs (15 vertex, 16 fragment) in 0.6 milliseconds
18:16:029 root         N[BOOT]: UI/EmuScreen.cpp:341 Loading /home/aaron/Downloads/ppsspp_old/Bleach_Heat_The_Soul_7_JPN_PSP-Caravan.iso...
18:21:688 Main         E[SCEIO]: HLE/sceIo.cpp:2450 Not a valid PGD file. Open as normal file.
18:21:689 Main         E[HLE]: HLE/HLE.cpp:374 sceNpDrmEdataSetupKey: Delaying a thread that's already waiting
18:21:720 Main         E[SCEIO]: HLE/sceIo.cpp:2450 Not a valid PGD file. Open as normal file.
18:21:720 Main         E[HLE]: HLE/HLE.cpp:374 sceNpDrmEdataSetupKey: Delaying a thread that's already waiting
18:55:805 Main         E[SCEKERNEL]: HLE/sceKernelThread.cpp:1340 80020198=sceKernelGetThreadExitStatus(0)
